### DISCUSSION TITLE
Unique value in a struct

### DISCUSSION THREAD

**Discussion by ahmed711** [cue-lang/cue#1833] [Q&A] [answered]
Hi all,
I am new to cue and I am trying to use cue to validate some json files, I created a schema cue file and I was wondering if the value can set to be unique to a certain field

For instance 

Employees: [...#Employee]
#Employee: {
  Id: int
  name: string 
}

Is there a way in cue to make sure no employees have the same Id ? 

Thanks

**Comment by mvdan** [comment, marked as answer]
You can use an object (map) to restrict that each Id only exists as one index to the Employees list: https://cuelang.org/play/?id=OGy7S-1wzhh#cue@export@cue

Also see https://github.com/cue-lang/cue/discussions/1213 for other ways to define a similar constraint.

**Reply by ahmed711** [reply]
This works perfectly, thank you
Can you direct me to understand better how this actually works ?

**Reply by mvdan** [reply]
We're essentially using one struct where we have one field per employee: the field name is the employee's ID, and the field value is their index within the employees list (this number must be unique, as you cannot have two elements in the same index in a list).

Due to how CUE works, you cannot have two fields with the same name but a different value, because they conflict - hence the error if there are any duplicate IDs in employees. See https://cuelang.org/docs/tutorials/tour/intro/duplicates/.

The docs page has lots for you to read if you want to learn more. The two key concepts here are structs and unification.

**Reply by ahmed711** [reply]
Thanks a lot @mvdan 
That was very helpful

building on what you mentioned I was expecting nested loops to work the same 
so for instance if we have the same 

Employees: [...#Employee]
#Employee: {
Id: int
name: string
company: string
}

and I don't mind having employees with the same id as long as they work for different companies 
but the id should not be a duplicate only if they work in the same company 

I thought something like this should do it but apparently not

#uniqPrefix: { 
    for emp in employees {
      let temp = emp.company
      for i, emp2 in employees{ 
        if emp2.company == temp 
          {
            "\(emp2.id)": i
          }
      }
    }
  }

**Reply by mvdan** [reply]
You should use the playground to share examples, others can easily reproduce the same behavior that you see :)

As for your second question, you can use exactly the same strategy I showed before with one loop, but using a field name that uses both the company string and ID integer. Using both as the field name string means that the combination of the two must be unique, like a joint primary key.

https://cuelang.org/play/?id=GkOQvwpFu7I#cue@export@cue
